File: Astrolabe.Evaluator/PartialEvaluator.cs
namespace Astrolabe.Evaluator;

/// <summary>
/// Partial evaluator that simplifies expressions by evaluating compile-time known values
/// while preserving runtime expressions.
/// </summary>
public static class PartialEvaluator
{
    /// <summary>
    /// Partially evaluates an expression given an environment with known values.
    /// Returns a simplified expression tree with compile-time values substituted.
    /// </summary>
    public static EvalExpr PartialEvaluate(this PartialEvalEnvironment env, EvalExpr expr)
    {
        return PartialEvaluateExpr(env, expr);
    }

    private static EvalExpr PartialEvaluateExpr(PartialEvalEnvironment env, EvalExpr expr)
    {
        return expr switch
        {
            // Already a value - return as-is
            ValueExpr v => v,

            // Variable - try to resolve
            VarExpr ve => PartialEvaluateVar(env, ve),

            // Property access - try to evaluate, keep as-is if it fails
            PropertyExpr pe => PartialEvaluateProperty(env, pe),

            // Function call - check for special cases
            CallExpr { Function: "?" } ce => PartialEvaluateTernary(env, ce),
            CallExpr { Function: "." } ce => PartialEvaluateFlatMap(env, ce),
            CallExpr ce => PartialEvaluateCall(env, ce),

            // Array literal
            ArrayExpr ae => PartialEvaluateArray(env, ae),

            // Let binding
            LetExpr le => PartialEvaluateLet(env, le),

            // Lambda - partially evaluate the body
            LambdaExpr lambda => PartialEvaluateLambda(env, lambda),

            _ => expr
        };
    }

    private static bool IsFullyEvaluated(EvalExpr expr)
    {
        return expr is ValueExpr;
    }

    private static EvalExpr PartialEvaluateVar(PartialEvalEnvironment env, VarExpr varExpr)
    {
        // Check if variable exists and get its value
        var value = env.TryGetVariable(varExpr.Name);
        if (value != null)
        {
            // Variable is defined with a compile-time value
            return value;
        }

        // Variable not found - keep as VarExpr (will be runtime value)
        return varExpr;
    }

    private static EvalExpr PartialEvaluateProperty(PartialEvalEnvironment env, PropertyExpr propExpr)
    {
        // Check if there's any data context at all
        if (!env.HasDataContext())
        {
            // No data context - this is a runtime field
            return propExpr;
        }

        // There's data context - evaluate the property access
        // (will return null/undefined if property doesn't exist, which is fine)
        var current = env.State.Current;

        if (current.Value is ObjectValue ov)
        {
            // Try to get the property value
            if (ov.Properties.TryGetValue(propExpr.Property, out var propValue))
            {
                return propValue;
            }
            // Property doesn't exist - return null ValueExpr
            return ValueExpr.Null;
        }

        // Current value is not an object - can't access property
        return ValueExpr.Null;
    }

    private static EvalExpr PartialEvaluateFlatMap(PartialEvalEnvironment env, CallExpr ce)
    {
        // ce.Args = [left, right] for the . operator
        var left = PartialEvaluateExpr(env, ce.Args[0]);
        var right = ce.Args[1];

        // If left is fully evaluated to an ObjectValue, update the environment's current context
        if (left is ValueExpr { Value: ObjectValue })
        {
            // Update environment with this value as current context
            var newEnv = (PartialEvalEnvironment)env.WithCurrent((ValueExpr)left);
            // Partially evaluate the right side with the new context
            return PartialEvaluateExpr(newEnv, right);
        }

        // If left is fully evaluated to an ArrayValue, we need to map over it
        if (left is ValueExpr { Value: ArrayValue av })
        {
            // For arrays, we'd need to map the partial evaluator over each element
            // For now, if we can't fully evaluate, keep as CallExpr
            var partialRight = PartialEvaluateExpr(env, right);
            if (IsFullyEvaluated(partialRight))
            {
                // Both sides evaluated, can try to execute
                try
                {
                    var reconstructed = new CallExpr(".", [left, partialRight], ce.Location);
                    var result = env.Evaluate(reconstructed);
                    return result.Value;
                }
                catch
                {
                    // Fall through to return CallExpr
                }
            }
            return new CallExpr(".", [left, partialRight], ce.Location);
        }

        // If left is null/undefined, return it
        if (left is ValueExpr ve && (ve.Value == null || ve == ValueExpr.Undefined))
        {
            return left;
        }

        // Otherwise, partially evaluate right and reconstruct
        var simplifiedRight = PartialEvaluateExpr(env, right);
        return new CallExpr(".", [left, simplifiedRight], ce.Location);
    }

    private static EvalExpr PartialEvaluateTernary(PartialEvalEnvironment env, CallExpr ce)
    {
        // ce.Args = [condition, trueBranch, falseBranch]
        var condition = PartialEvaluateExpr(env, ce.Args[0]);

        if (condition is ValueExpr condValue)
        {
            // Condition is known - pick the appropriate branch
            // Convert to bool - treat null/undefined/false as false, everything else as true
            bool isTrue = condValue.Value switch
            {
                null => false,
                bool b => b,
                _ => condValue != ValueExpr.Undefined
            };

            var selectedBranch = isTrue ? ce.Args[1] : ce.Args[2];
            return PartialEvaluateExpr(env, selectedBranch);
        }
        else
        {
            // Condition is unknown - simplify all three parts
            var trueBranch = PartialEvaluateExpr(env, ce.Args[1]);
            var falseBranch = PartialEvaluateExpr(env, ce.Args[2]);
            return new CallExpr("?", [condition, trueBranch, falseBranch], ce.Location);
        }
    }

    private static EvalExpr PartialEvaluateCall(PartialEvalEnvironment env, CallExpr ce)
    {
        // Partially evaluate all arguments
        var partialArgs = ce.Args.Select(arg => PartialEvaluateExpr(env, arg)).ToList();

        // If all arguments are fully evaluated, try to execute the function
        if (partialArgs.All(IsFullyEvaluated))
        {
            // All arguments are ValueExpr - we can call the function handler
            var reconstructed = new CallExpr(ce.Function, partialArgs, ce.Location);

            // Try to evaluate using the standard evaluator
            // The PartialEvalEnvironment.Evaluate accepts ValueExpr only, which is what we have
            try
            {
                var result = env.Evaluate(reconstructed);
                return result.Value;
            }
            catch
            {
                // Evaluation failed - keep as call expression
                // This can happen if the function doesn't exist or errors during execution
            }
        }

        // Reconstruct with simplified arguments
        return new CallExpr(ce.Function, partialArgs, ce.Location);
    }

    private static EvalExpr PartialEvaluateArray(PartialEvalEnvironment env, ArrayExpr ae)
    {
        var partialElements = ae.Values.Select(v => PartialEvaluateExpr(env, v)).ToList();

        // If all elements are fully evaluated, create an ArrayValue
        if (partialElements.All(IsFullyEvaluated))
        {
            var values = partialElements.Cast<ValueExpr>().ToList();
            return new ValueExpr(new ArrayValue(values));
        }

        // Otherwise, keep as ArrayExpr with simplified elements
        return new ArrayExpr(partialElements, ae.Location);
    }

    private static EvalExpr PartialEvaluateLet(PartialEvalEnvironment env, LetExpr le)
    {
        // Partially evaluate each binding and extend the environment
        var newEnv = env;
        var newVars = new List<(VarExpr, EvalExpr)>();

        foreach (var (varExpr, bindingExpr) in le.Vars)
        {
            var partialBinding = PartialEvaluateExpr(newEnv, bindingExpr);

            if (partialBinding is ValueExpr)
            {
                // Binding is fully evaluated - add to environment for future lookups
                newEnv = (PartialEvalEnvironment)newEnv.WithVariable(varExpr.Name, partialBinding);
            }
            else
            {
                // Binding has runtime dependencies - keep in let expression
                newVars.Add((varExpr, partialBinding));
            }
        }

        // Partially evaluate the body with the extended environment
        var partialBody = PartialEvaluateExpr(newEnv, le.In);

        // If no variables remain, just return the body
        if (newVars.Count == 0)
        {
            return partialBody;
        }

        // Reconstruct let expression with remaining variables
        return new LetExpr(newVars, partialBody, le.Location);
    }

    private static EvalExpr PartialEvaluateLambda(PartialEvalEnvironment env, LambdaExpr lambda)
    {
        // Lambda parameters are unknown at compile-time, but we can still partially
        // evaluate the body using the compile-time values available in the environment

        // The lambda parameter ($x, $i, etc.) is an unknown runtime value
        // We don't add it to the environment, so references to it will remain as VarExpr

        // Partially evaluate the lambda body
        var partialBody = PartialEvaluateExpr(env, lambda.Value);

        // If the body hasn't changed, return the original lambda
        if (ReferenceEquals(partialBody, lambda.Value))
        {
            return lambda;
        }

        // Reconstruct lambda with simplified body
        return new LambdaExpr(lambda.Variable, partialBody, lambda.Location);
    }
}

File: astrolabe-evaluator/src/partialEvaluator.ts
import {
  EvalEnv,
  EvalEnvState,
  EvalExpr,
  ValueExpr,
  VarExpr,
  PropertyExpr,
  CallExpr,
  ArrayExpr,
  LetExpr,
  LambdaExpr,
  EnvValue,
  valueExpr,
  lookupVar,
} from "./ast";
import { defaultEvaluate } from "./evaluate";

/**
 * Specialized environment for partial evaluation that distinguishes between
 * compile-time values (variables) and runtime fields (properties).
 */
export class PartialEvalEnvironment extends EvalEnv {
  constructor(public state: EvalEnvState) {
    super();
  }

  /**
   * Override evaluateExpr - main implementation for partial evaluation.
   * Handles all expression types, returning partially-evaluated results.
   */
  evaluateExpr(expr: EvalExpr): EnvValue<EvalExpr> {
    switch (expr.type) {
      case "value":
        return [this, expr];

      case "var": {
        // Try to substitute with compile-time value if known
        const varValue = this.getVariable(expr.variable);
        return varValue
          ? this.evaluateExpr(varValue)  // Recursively evaluate
          : [this, expr];                 // Keep as runtime variable
      }

      case "property": {
        // Only evaluate if we have data context
        if (this.hasDataContext()) {
          return [this, this.state.data.getProperty(this.state.current, expr.property)];
        }
        return [this, expr];  // Keep as runtime property
      }

      case "call": {
        // If function exists, call defaultEvaluate; else return CallExpr unchanged
        const funcValue = this.getVariable(expr.function);
        if (funcValue && funcValue.function) {
          const [env, result] = defaultEvaluate(this, expr);
          return [env, result];  // result is already EvalExpr
        }
        return [this, expr];  // Unknown function - keep as partial CallExpr
      }

      case "let": {
        // Evaluate all bindings and add to environment
        let newEnv: PartialEvalEnvironment = this;
        const bindings: Record<string, EvalExpr> = {};

        for (const [varExpr, binding] of expr.variables) {
          const [nextEnv, evaluated] = newEnv.evaluateExpr(binding);
          bindings[varExpr.variable] = evaluated;
          newEnv = nextEnv.withVariable(varExpr.variable, evaluated) as PartialEvalEnvironment;
        }

        // Evaluate body with extended environment
        const [finalEnv, body] = newEnv.evaluateExpr(expr.expr);

        // Collect free variables in the result
        const freeVars = collectFreeVars(body);
        const neededBindings = Object.entries(bindings)
          .filter(([name, val]) => freeVars.has(name) && val.type !== 'value')
          .map(([name, val]): [VarExpr, EvalExpr] => [{ type: 'var', variable: name }, val]);

        // Return body alone if no partial bindings needed, else wrap in LetExpr
        return neededBindings.length === 0
          ? [finalEnv, body]
          : [finalEnv, { type: 'let', variables: neededBindings, expr: body }];
      }

      case "array": {
        // Evaluate all elements
        const [envAfterArray, elements] = mapAllEnv(this, expr.values, (env, val) => env.evaluateExpr(val));

        // If all elements are ValueExpr, return a ValueExpr array
        if (elements.every(e => e.type === 'value')) {
          return [envAfterArray, valueExpr(elements as ValueExpr[])];
        }

        // Otherwise keep as ArrayExpr
        return [envAfterArray, { type: 'array', values: elements }];
      }

      case "lambda":
        // Keep lambda as-is (parameters are unknown at compile-time)
        return [this, expr];

      default:
        throw new Error(
          `PartialEvalEnvironment.evaluateExpr cannot handle expression type: ${(expr as any).type}`,
        );
    }
  }

  /**
   * Override evaluate() to throw error - partial eval should use evaluateExpr.
   */
  evaluate(expr: EvalExpr): EnvValue<ValueExpr> {
    throw new Error(
      "Use evaluateExpr() for partial evaluation. " +
        "evaluate() is only for standard full evaluation.",
    );
  }

  /**
   * Check if a variable is defined in the environment (compile-time value).
   */
  hasVariable(name: string): boolean {
    let current: EvalEnvState | undefined = this.state;
    while (current) {
      if (name in current.localVars) return true;
      current = current.parent;
    }
    return false;
  }

  /**
   * Check if there is any data context (for PropertyExpr evaluation).
   * Returns false if current value is undefined or null (no data context).
   */
  hasDataContext(): boolean {
    const current = this.state.current;
    return current.value !== undefined && current.value !== null;
  }

  /**
   * Try to get a variable value if it exists.
   * Returns undefined if the variable is not found.
   */
  tryGetVariable(name: string): ValueExpr | undefined {
    return lookupVar(this.state, name);
  }

  protected newEnv(newState: EvalEnvState): EvalEnv {
    return new PartialEvalEnvironment(newState);
  }

  /**
   * Override withVariable to not evaluate the variable value.
   * In partial evaluation, we want to keep the raw EvalExpr until we know if it's fully evaluatable.
   */
  withVariable(name: string, expr: EvalExpr): EvalEnv {
    // For partial evaluation, if expr is a ValueExpr, use it directly
    // Otherwise, we can't add it as a variable (would need to be partially evaluated first)
    if (expr.type === "value") {
      return this.newEnv({
        data: this.state.data,
        current: this.state.current,
        compare: this.state.compare,
        localVars: { [name]: expr },
        parent: this.state,
        errors: this.state.errors,
      });
    }

    // If it's not a value, we can't add it to variables yet
    // This should not happen in partial evaluation since we partially evaluate bindings first
    throw new Error(
      "PartialEvalEnvironment.withVariable only accepts ValueExpr",
    );
  }

  withVariables(vars: [string, EvalExpr][]): EvalEnv {
    if (vars.length === 0) {
      return this;
    }

    if (vars.length === 1) {
      return this.withVariable(vars[0][0], vars[0][1]);
    }

    // Build up variables
    const newVars: Record<string, ValueExpr> = {};
    for (const [name, expr] of vars) {
      if (expr.type === "value") {
        newVars[name] = expr;
      } else {
        throw new Error(
          "PartialEvalEnvironment.withVariables only accepts ValueExpr",
        );
      }
    }

    return this.newEnv({
      data: this.state.data,
      current: this.state.current,
      compare: this.state.compare,
      localVars: newVars,
      parent: this.state,
      errors: this.state.errors,
    });
  }

  withCurrent(current: ValueExpr): EvalEnv {
    return this.newEnv({ ...this.state, current });
  }

  withError(error: string): EvalEnv {
    return this.newEnv({
      ...this.state,
      errors: [...this.state.errors, error],
    });
  }

  getVariable(name: string): ValueExpr | undefined {
    return lookupVar(this.state, name);
  }

  compare(v1: unknown, v2: unknown): number {
    return this.state.compare(v1, v2);
  }

  get current() {
    return this.state.current;
  }

  get errors() {
    return this.state.errors;
  }

  get data() {
    return this.state.data;
  }
}

/**
 * Collect free variables in an expression (variables not bound by lambda/let in that expression).
 * Handles scoping correctly - lambda and let bindings shadow outer variables.
 */
function collectFreeVars(expr: EvalExpr, boundVars: Set<string> = new Set()): Set<string> {
  const freeVars = new Set<string>();

  function collect(e: EvalExpr, bound: Set<string>) {
    switch (e.type) {
      case "var":
        if (!bound.has(e.variable)) {
          freeVars.add(e.variable);
        }
        break;

      case "lambda": {
        // Lambda parameter shadows outer variables
        const innerBound = new Set(bound);
        innerBound.add(e.variable);
        collect(e.expr, innerBound);
        break;
      }

      case "let": {
        // Let bindings shadow outer variables
        const letBound = new Set(bound);
        for (const [v, _] of e.variables) {
          letBound.add(v.variable);
        }

        // Binding expressions see outer scope
        for (const [_, binding] of e.variables) {
          collect(binding, bound);
        }

        // Body sees let-bound variables
        collect(e.expr, letBound);
        break;
      }

      case "call":
        for (const arg of e.args) {
          collect(arg, bound);
        }
        break;

      case "array":
        for (const val of e.values) {
          collect(val, bound);
        }
        break;

      case "value":
        if (Array.isArray(e.value)) {
          for (const val of e.value) {
            collect(val, bound);
          }
        } else if (typeof e.value === "object" && e.value !== null) {
          for (const val of Object.values(e.value)) {
            if (val && typeof val === "object" && "type" in val) {
              collect(val as EvalExpr, bound);
            }
          }
        }
        break;

      case "property":
        // No variables to collect
        break;
    }
  }

  collect(expr, boundVars);
  return freeVars;
}

/**
 * Partially evaluates an expression given an environment with known values.
 * Returns a simplified expression tree with compile-time values substituted.
 */
export function partialEvaluate(
  env: PartialEvalEnvironment,
  expr: EvalExpr,
): EvalExpr {
  const [_, result] = env.evaluateExpr(expr);
  return result;
}
